### ISSUE TITLE
Evaluate MJX-inspired execution architecture for Learned Policy self-play and evaluation

### ISSUE THREAD

**Issue by lisbun** [lisbun/lisjong-project#41]
## Background

MJX was reviewed as a reference implementation for large-scale Mahjong simulation and learning workflows.

The main conclusion is **not** to adopt MJX as a new canonical backend or to migrate away from the current RiichiEnv-based evaluation path. The useful part is its execution architecture around learned agents and large-scale match generation.

The following MJX ideas appear relevant to the lisjong ecosystem:

- parallel game execution
- batched policy inference across multiple concurrent games
- separation between game execution and inference service
- self-contained/stateless per-decision observation contract
- duplicate game scheduling to reduce seat / wall variance
- reconstruction of historical `(observation, legal actions, action)` decisions from game state / replay
- reference-corpus and differential validation against an independent mature implementation

Current lisjong design already has stronger contracts in several areas and should retain them:

- `PolicyInput` / `DecisionContext` remain lisjong-owned, engine-independent decision contracts
- `InternalAction` remains the canonical semantic action model
- `lisjong-action-vocabulary-1` and its legal-mask / canonical-resolution semantics remain the model-facing action contract
- feature / tensor schemas remain separately versioned experiment representations rather than engine-owned observation features
- RiichiEnv remains the current canonical Arena evaluation backend unless a separate bounded decision changes that

This Issue captures the architectural lessons and decides which ones should become bounded implementation work.

## Goal

Define a lisjong-owned / Arena-owned architecture for high-throughput Learned Policy execution that can support future behavior cloning, self-play, and RL experiments **without coupling Policy semantics to a particular simulator or inference transport**.

### 1. Batch inference seam

Evaluate an Arena-owned execution seam conceptually similar to:

```text
ParallelGameWorker × N
        |
        | DecisionRequest
        v
DynamicBatcher
  - PolicyInput
  - legal action mask
  - decision identity
        |
        v
Learned Policy model batch inference
        |
        v
resolve_legal_action()
```

Requirements to consider:

- existing scalar `Policy.choose_action()` contract must remain valid for heuristic / non-batched policies
- batching is an execution optimization, not a new Policy semantic contract
- model inference must consume lisjong-owned / versioned feature and action contracts rather than simulator-native types
- preserve deterministic result identity where required

### 2. Parallel self-play architecture

Define the boundary between:

- game workers
- Policy / model inference
- result / trace collection
- training dataset production

Start with same-process execution as the default design. Distributed transport such as gRPC is **not** required unless later measurements justify it.

### 3. Reproducible duplicate scheduling

Evaluate a deterministic duplicate schedule inspired by MJX, but with stronger reproducibility requirements.

A duplicate group should be able to identify at least:

- master / game seed
- seat permutation
- participating Policy identities / revisions
- backend revision / identity
- relevant action-vocabulary / feature-schema identities when a learned Policy is involved

Do not copy MJX's non-reproducible random duplicate seed generator as-is.

### 4. Replay -> decision dataset reconstruction

Evaluate a canonical offline path from an Arena game trace / replay source to decision-level records such as:

```text
DecisionRecord
  - PolicyInput
  - legal action mask
  - selected action index
  - canonical InternalAction
  - game identity
  - decision identity
  - seat
```

The design must preserve the information boundary:

- full hidden state may be available to offline evaluation / supervision
- hidden ground truth must not leak into runtime `PolicyInput`

Consider whether this can become a common substrate for:

- behavior cloning datasets
- self-play datasets
- HandBelief offline ground-truth evaluation
- decision-level debugging / analysis

### 5. Independent reference / differential validation

Assess MJX and/or Tenhou-derived corpus cases as an independent reference source for difficult rule / transition cases.

Potential targets include:

- furiten
- riichi lifecycle
- chi / pon / kan transitions
- ankan / kakan / daiminkan
- chankan
- multiple ron
- abortive draws
- nagashi mangan
- score settlement

Agreement with MJX must **not** be treated as proof of correctness. The value is disagreement detection and test-case discovery.

## Explicit non-goals

- replacing RiichiEnv with MJX
- adding MJX as a production runtime dependency
- adopting MJX's protobuf schema as a lisjong domain contract
- replacing the current 802-action lisjong model vocabulary with MJX's 181-action encoding
- moving feature semantics into the simulator
- implementing distributed gRPC inference before a measured need exists
- premature C++ / native optimization without current bottleneck evidence
- redesigning `PolicyInput`, `InternalAction`, or current action vocabulary solely to resemble MJX

## Questions to resolve

1. What is the smallest stable request / response boundary needed for batched inference?
2. Should batching live entirely in `lisjong-arena`, with `lisjong` owning only scalar semantic contracts and model logic?
3. What trace artifact is sufficient to reconstruct canonical decision records without relying on simulator-specific mutable state?
4. What deterministic duplicate schedule should Arena standardize for Learned Policy comparison?
5. Which identities must be persisted so generated training / evaluation corpora are reproducible?
6. Which parts should be implemented before the first substantial self-play dataset is generated?
7. Does current throughput measurement justify any backend work now, or should backend replacement remain deferred?

## Expected outcome

This Issue should finish with an architectural decision, not a large implementation PR.

Expected deliverables:

- documented recommended architecture and ownership boundaries
- explicit KEEP / ADOPT / DEFER / REJECT decisions for the MJX-inspired ideas above
- concrete reproducibility requirements for parallel / duplicate execution
- definition of the minimal batch-inference seam, if adopted
- definition of the minimal replay-to-decision record, if adopted
- identification of bounded child Issues only where implementation is now justified

## Proposed priority

Recommended order if the review confirms value:

1. batch-inference execution seam
2. parallel self-play runner around that seam
3. deterministic duplicate scheduling
4. replay-to-decision dataset reconstruction
5. reference / differential-validation corpus work
6. distributed inference only after local batching is measured

## Decision principle

Use MJX as a **reference implementation and source of architectural ideas**, not as the owner of lisjong semantics.

The intended direction is:

```text
lisjong
  PolicyInput / InternalAction / action vocabulary / model semantics
                    |
                    v
lisjong-arena
  parallel execution / batching / replay / evaluation / artifacts
            |                       |
        RiichiEnv              future backend?
                                  MJX spike?
```

Backends should remain replaceable behind stable lisjong contracts.

**Comment by lisbun** [comment]
## Architecture review complete — MJX ideas classified

Current Learned Policy state and Arena architecture were re-reviewed after this Issue was opened.

Relevant current context:

- the original minimal BC candidate completed serving integration but failed the bounded strength screen with `DO NOT ADVANCE`
- the successor Learned Policy roadmap selected `OFFLINE VALUE / Q EXPERIMENT RECOMMENDED`
- the current bounded Offline Q experiment uses the existing `PolicyInput` / feature / action / serving contracts and does not require broad self-play infrastructure
- Arena already owns game-level local process parallelism for AABB / ABBB
- current `LocalGameInspection` already provides a concrete same-process source of `GameTrace` + step-scoped `PolicyInput` + validated `DecisionTrace`

Therefore this review does **not** justify implementing a broad MJX-like execution framework now.

### Final classification

| Idea | Decision |
| --- | --- |
| lisjong-owned self-contained `PolicyInput` / decision contract | **KEEP** |
| scalar `Policy.choose_action()` semantic contract | **KEEP** |
| `InternalAction` + `lisjong-action-vocabulary-1` + legal-mask / canonical-resolution semantics | **KEEP** |
| RiichiEnv as current canonical Arena evaluation backend | **KEEP** |
| existing Arena game-level process parallelism | **KEEP** |
| Arena-owned batch-inference execution seam | **ADOPT ARCHITECTURE / DEFER IMPLEMENTATION** |
| parallel self-play runner around the batch seam | **DEFER** |
| deterministic duplicate scheduling requirements | **ADOPT REQUIREMENTS / DEFER GENERIC IMPLEMENTATION** |
| minimal replay/inspection -> decision-record semantics | **ADOPT LOGICAL CONTRACT / DEFER GENERIC PERSISTENCE** |
| MJX / independent implementation differential validation | **ADOPT AS DIAGNOSTIC STRATEGY / DEFER AUTOMATION** |
| MJX as canonical backend or production dependency | **REJECT** |
| MJX protobuf as lisjong domain contract | **REJECT** |
| replacing the 802-action vocabulary with MJX's action encoding | **REJECT** |
| distributed gRPC inference | **DEFER UNTIL MEASURED NEED** |
| native/C++ optimization | **DEFER UNTIL PROFILED NEED** |

## Ownership decision

The execution boundary remains:

```text
lisjong
    PolicyInput / DecisionContext
    InternalAction
    action vocabulary semantics
    model / Policy semantics where production-owned

                 |
                 v

lisjong-arena
    concrete game execution
    parallel scheduling
    optional inference batching
    trace / inspection composition
    evaluation
    experiment datasets / artifacts
```

Batching is an **Arena execution optimization**, not a new Policy semantic contract.

The existing scalar Policy contract remains valid for heuristic and non-batched Policies.

## Minimal future batch-inference seam

If profiling later justifies batching, the smallest intended seam is conceptually:

```text
DecisionRequest
    decision_identity
    inference_identity
    PolicyInput
    legal_action_mask

        ↓

Arena-owned batch coordinator

        ↓

versioned feature encoder
learned model batch inference

        ↓

DecisionResponse
    decision_identity
    selected_action_index
```

The caller retains the original decision context and performs canonical:

```text
selected action index
    -> resolve_legal_action(...)
    -> InternalAction
    -> existing validation / execution boundary
```

No simulator-native Observation / Action becomes part of the model contract.

`decision_identity` must be deterministic and independent of batching order or worker completion order. At minimum it must distinguish:

```text
game identity
decision ordinal
actor seat
```

The exact concurrency primitive is intentionally **not locked** here. Threads, cooperative execution, process IPC, or another mechanism should be selected only after a concrete batching consumer and measurement exist.

## Parallel self-play decision

No dedicated self-play runner is implemented by this Issue.

Future execution should keep these responsibilities separable:

```text
game scheduling
game execution
Policy/model inference
objective trace / result collection
decision-dataset construction
training
```

Same-machine execution remains the default first design. Distributed transport is not introduced until local measurements justify it.

The current Offline Value/Q experiment must not wait for this infrastructure.

## Duplicate scheduling requirements

Future learned-policy comparisons / self-play may standardize a duplicate-group contract, but the generic scheduler is deferred.

A reproducible duplicate group must persist enough information to identify at least:

```text
schedule semantics/version
master/game seed or exact ordered game seeds
seat permutation
participant Policy identities/revisions
learned artifact identity where applicable
backend + rules identity/revision
feature schema identity where applicable
action vocabulary identity where applicable
```

The schedule must be reproducible without relying on process-global RNG state or a non-recorded random seed generator.

Existing Arena fixed-seed + deterministic seat-rotation protocols remain valid and should be reused rather than replaced.

## Minimal decision-record semantics

A future reusable decision substrate should logically be able to represent:

```text
DecisionRecord
    game_identity
    decision_identity
    actor_seat

    PolicyInput
    legal_action_mask
    selected_action_index
    canonical InternalAction

    feature_schema_identity
    action_vocabulary_identity
```

This is a logical contract, **not a requirement to introduce a generic serializer or canonical GameRecord now**.

Current `GameTrace` remains objective execution observation and should not absorb Policy-visible state, labels, hidden truth, rewards, or model-specific training semantics.

Purpose-specific consumers may enrich the core decision data separately:

```text
DecisionRecord
    + Offline-Q transition / reward
    + behavior-cloning label
    + HandBelief hidden ground truth
    + diagnostic metadata
```

The runtime information boundary remains:

```text
player-safe PolicyInput
        -> runtime model / Policy

omniscient hidden truth / future reward
        -> explicit offline supervision only
```

No hidden ground truth may leak into runtime `PolicyInput`.

## Differential validation

MJX and other mature independent implementations remain useful for:

```text
disagreement detection
edge-case discovery
curated regression-case generation
```

They are not correctness authorities.

No production MJX dependency or broad differential-testing framework is introduced until a concrete rule/transition validation consumer justifies it.

## Implementation gate

No new batcher / self-play / generic replay child Issue is required from this review.

Implementation becomes justified when at least one concrete consumer exists and measurement shows that the current execution path is a material limiter, for example:

- substantial self-play or online-RL generation becomes the selected next step
- learned-model inference becomes a meaningful share of total game-generation runtime
- process-per-game learned execution causes measured model duplication / inference overhead
- multiple consumers require the same persisted decision-record substrate
- a concrete duplicate-evaluation campaign cannot be expressed cleanly by existing Arena protocols

Until then, the current bounded Learned Policy / HandBelief experiments should use their purpose-specific Arena paths.

## Final decision

```text
MJX EXECUTION IDEAS ADOPTED SELECTIVELY
IMPLEMENTATION DEFERRED UNTIL CONCRETE MEASURED NEED
```

The useful MJX lesson is the separation of game execution, inference, and data production — not adoption of MJX's backend, protobuf, action representation, or distributed architecture.

This Issue can close with no implementation child required at this time.
